panel/applets/icon-tasklist/WindowMenu.vala

```diff
@@ -0,0 +1,161 @@
+     * need this menu, in order for the destructor of WindowMenu to be called.
+     * Perhaps there is a reference cycle between Wnck.Window and ActionMenu.
+     */
+    public void clear()
+    {
+        menu.destroy();
+    }
+
+    public void update_app_info(GLib.DesktopAppInfo? app_info)
+    {
+        this.app_info = app_info;
+    }
+
+    public void show(uint event_button, uint32 timestamp)
+    {
+        menu.popup(null, null, null, event_button, timestamp);
+    }
+
+    public void enable_pinning_option()
+    {
+        unpinnage.hide();
+        pinnage.show();
+        sep_item.show();
+
+        pinning_enabled = true;
+        unpinning_enabled = false;
+    }
+
+    public void enable_unpinning_option()
+    {
+        unpinnage.show();
+        pinnage.hide();
+        sep_item.show();
+
+        pinning_enabled = false;
+        unpinning_enabled = true;
+    }
+
+    public void disable_all_pinning_options()
+    {
+        unpinnage.hide();
+        pinnage.hide();
+        sep_item.hide();
+
+        pinning_enabled = false;
+        unpinning_enabled = false;
+    }
+
+    private void build_application_menu(Wnck.Window window, GLib.DesktopAppInfo? app_info)
+    {
+        // Actions menu
+        menu = new Wnck.ActionMenu(window);
+
+        var sep = new Gtk.SeparatorMenuItem();
+        menu.append(sep);
+        sep_item = sep;
+        close_all_windows_opt = new Gtk.MenuItem.with_label(_("Close All"));
+        pinnage = new Gtk.MenuItem.with_label(_("Pin to panel"));
+        unpinnage = new Gtk.MenuItem.with_label(_("Unpin from panel"));
+        sep.show();
+        close_all_windows_opt.show();
+        menu.append(close_all_windows_opt);
+        menu.append(pinnage);
+        menu.append(unpinnage);
+
+        close_all_windows_opt.activate.connect(()=> {
+            application.close_all_windows();
+        });
+        /* Handle running instance pin/unpin */
+        pinnage.activate.connect(()=> {
+            assert(pinning_enabled);
+            DesktopHelper.set_pinned(settings, app_info, true);
+        });
+
+        unpinnage.activate.connect(()=> {
+            assert(unpinning_enabled);
+            DesktopHelper.set_pinned(settings, app_info, false);
+        });
+
+        if (app_info != null) {
+            // Desktop app actions =)
+            unowned string[] actions = app_info.list_actions();
+            if (actions.length == 0) {
+                return;
+            }
+            sep = new Gtk.SeparatorMenuItem();
+            menu.append(sep);
+            sep.show_all();
+            foreach (var action in actions) {
+                var display_name = app_info.get_action_name(action);
+                var item = new Gtk.MenuItem.with_label(display_name);
+                item.set_data("__aname", action);
+                item.activate.connect(()=> {
+                    string? act = item.get_data("__aname");
```

Comment:
Would be nice if we could move away from lambdas tbh, Vala lambdas are a heap of shite at best and damn near impossible to debug